### PR TITLE
Mauro/ipc cli serv qos getters

### PR DESCRIPTION
**Pull request by mauropasse**

### CODE REVIEW

File: rclcpp/include/rclcpp/client.hpp [hunk 1]
```diff
@@ -581,7 +640,7 @@ class Client : public ClientBase
     PromiseWithRequest promise;
     auto shared_future = promise.get_future().share();
     auto req_id = async_send_request_impl(
-      *request,
+      request,
```
Comment (alsora) [inline rclcpp/include/rclcpp/client.hpp:643]: std::move missing here?

File: rclcpp/include/rclcpp/client.hpp [hunk 2]
```diff
@@ -694,10 +753,28 @@ class Client : public ClientBase
...
   {
+    if (use_intra_process_) {
+      auto ipm = weak_ipm_.lock();
+      if (!ipm) {
+        throw std::runtime_error(
+                "intra process send called after destruction of intra process manager");
+      }
+      bool intra_process_server_available = ipm->service_is_available(intra_process_client_id_);
+
+      if (intra_process_server_available) {
```
Comment (alsora) [inline rclcpp/include/rclcpp/client.hpp:769]: why if the server is not available we fall back to DDS comm?

File: rclcpp/include/rclcpp/client.hpp [hunk 3]
```diff
@@ -694,10 +753,28 @@ class Client : public ClientBase
...
+          intra_process_client_id_,
+          std::make_pair(std::move(request), std::move(value)));
+        return sequence_number_++;
+      }
+    }
+
+    // Send inter-process request
     int64_t sequence_number;
-    rcl_ret_t ret = rcl_send_request(get_client_handle().get(), &request, &sequence_number);
+    rcl_ret_t ret = rcl_send_request(get_client_handle().get(), &(*request), &sequence_number);
```
Comment (alsora) [inline rclcpp/include/rclcpp/client.hpp]: can `&(*request)` become `request.get()` ?

File: rclcpp/include/rclcpp/client.hpp (hunk 1 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/client.hpp:643]: `request` is also used in line 646, if I move it, would be invalid

File: rclcpp/include/rclcpp/client.hpp [hunk 4]
```diff
@@ -735,6 +812,11 @@ class Client : public ClientBase
       CallbackInfoVariant>>
   pending_requests_;
   std::mutex pending_requests_mutex_;
+
+private:
+  using ClientIntraProcessT = rclcpp::experimental::ClientIntraProcess<ServiceT>;
+  std::shared_ptr<ClientIntraProcessT> client_intra_process_;
+  std::atomic_uint sequence_number_{1};
```
Comment (alsora) [inline rclcpp/include/rclcpp/client.hpp]: should this be named "ipc_sequence_number_" ?

File: rclcpp/include/rclcpp/create_client.hpp [hunk 5]
```diff
@@ -35,7 +35,8 @@ create_client(
   std::shared_ptr<node_interfaces::NodeServicesInterface> node_services,
   const std::string & service_name,
   const rmw_qos_profile_t & qos_profile,
-  rclcpp::CallbackGroup::SharedPtr group)
+  rclcpp::CallbackGroup::SharedPtr group,
+  rclcpp::IntraProcessSetting ip_setting)
```
Comment (alsora) [inline rclcpp/include/rclcpp/create_client.hpp]: let's use ipc_setting for the variable name... ip_setting seems the network IP

File: rclcpp/include/rclcpp/client.hpp (hunk 2 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/client.hpp:769]: Because the server might be available in another process, or created with IPC OFF.

Comment (alsora) [reply rclcpp/include/rclcpp/client.hpp:769]: Good point.
We should add a comment about it.

File: rclcpp/include/rclcpp/create_client.hpp (hunk 5 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/create_client.hpp]: Done in 3e4d9c65

File: rclcpp/include/rclcpp/client.hpp (hunk 4 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/client.hpp]: Done in 3e4d9c65

File: rclcpp/include/rclcpp/client.hpp (hunk 3 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/client.hpp]: Yes, done in 3e4d9c65

File: rclcpp/include/rclcpp/client.hpp (hunk 2 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/client.hpp:769]: Done in 3e4d9c65

**Pull request closed without merging**